src/api/wayfire/unstable/wlr-text-input-v3-popup.hpp

```diff
@@ -7,11 +7,20 @@
 
 namespace wf
 {
+class text_input_v3_popup;
+
 class text_input_v3_im_relay_interface_t
 {
+  protected:
```

Comment:
This seems like the wrong place to store information about active popups.

Note that popups already store a pointer to the input method relay. So I think here we can use a signal (make the relay derive from signal::provider_t) and then emit a signal on commit. Any popups currently active can connect to this signal and update their internal state, and we don't have to maintain a list of popups.

Reply:
Signal is added. That list of popups needs to stay there for input method v2 because otherwise the popup would be destroyed as nothing references it.